File: apps/react-native/src/domain/repositories.ts
// Repository ports: the boundary between business logic and the outside world. Use cases
// depend on these interfaces, never on the HTTP client.

import { Outcome } from '../core/core';
import { EventDetail, EventPage, Order, QueueToken, Reservation, TokenPair } from './models';

export interface EventRepository {
  listEvents(cursor?: string, limit?: number): Promise<Outcome<EventPage>>;
  getEvent(id: string): Promise<Outcome<EventDetail>>;
}

export interface QueueRepository {
  join(eventId: string): Promise<Outcome<QueueToken>>;
  status(eventId: string): Promise<Outcome<QueueToken>>;
}

export interface ReservationRepository {
  /** Creates a hold. `idempotencyKey` makes a retried create a no-op. */
  create(sectorId: string, quantity: number, idempotencyKey: string): Promise<Outcome<Reservation>>;
  release(id: string): Promise<Outcome<void>>;
}

export interface OrderRepository {
  /** Creates an order (async payment). `idempotencyKey` protects retries. */
  create(reservationId: string, idempotencyKey: string): Promise<Outcome<Order>>;
  get(id: string): Promise<Outcome<Order>>;
}

/** Produces client-side idempotency keys. A port so tests inject a deterministic sequence. */
export interface IdempotencyKeyFactory {
  newKey(): string;
}

/** Authenticates and rotates tokens. Login proves identity with credentials; refresh proves
 * it with the refresh token itself. */
export interface AuthRepository {
  login(email: string, password: string): Promise<Outcome<TokenPair>>;
  /** Exchanges the refresh token for a new, rotated pair. A failure means the session is over. */
  refresh(refreshToken: string): Promise<Outcome<TokenPair>>;
}

/** Persists the token pair: access token in memory, refresh token in the platform secure
 * store. InMemoryTokenStore is the test/demo implementation. */
export interface TokenStore {
  current(): TokenPair | null;
  save(tokens: TokenPair): void;
  clear(): void;
}

File: apps/flutter/lib/domain/repositories.dart
/// Repository ports: the boundary between business logic and the outside world. Use cases
/// depend on these, never on dio. Each returns a validated domain model in an [Outcome];
/// invalid data becomes a typed failure, never an exception that escapes.
library;

import '../core/core.dart';
import 'models.dart';

abstract interface class EventRepository {
  Future<Outcome<EventPage>> listEvents({String? cursor, int? limit});
  Future<Outcome<EventDetail>> getEvent(String id);
}

abstract interface class QueueRepository {
  Future<Outcome<QueueToken>> join(String eventId);
  Future<Outcome<QueueToken>> status(String eventId);
}

abstract interface class ReservationRepository {
  /// Creates a hold. [idempotencyKey] makes a retried create a no-op, so a double tap
  /// cannot create two reservations.
  Future<Outcome<Reservation>> create(String sectorId, int quantity, String idempotencyKey);
  Future<Outcome<void>> release(String id);
}

abstract interface class OrderRepository {
  /// Creates an order, triggering async payment. [idempotencyKey] protects retries after a
  /// network drop or unknown-outcome timeout.
  Future<Outcome<Order>> create(String reservationId, String idempotencyKey);
  Future<Outcome<Order>> get(String id);
}

/// Produces client-side idempotency keys. A port so tests inject a deterministic sequence.
abstract interface class IdempotencyKeyFactory {
  String newKey();
}
